File: sdks/csharp/Windjammer.SDK/Math.cs
using System;

namespace Windjammer.SDK;

/// <summary>
/// 2D vector.
/// </summary>
public struct Vector2 : IEquatable<Vector2>
{
    /// <summary>X component.</summary>
    public float X;
    
    /// <summary>Y component.</summary>
    public float Y;

    /// <summary>
    /// Creates a new 2D vector.
    /// </summary>
    /// <param name="x">X component.</param>
    /// <param name="y">Y component.</param>
    public Vector2(float x, float y)
    {
        X = x;
        Y = y;
    }

    /// <summary>Adds two vectors.</summary>
    public static Vector2 operator +(Vector2 a, Vector2 b) => new(a.X + b.X, a.Y + b.Y);
    
    /// <summary>Subtracts two vectors.</summary>
    public static Vector2 operator -(Vector2 a, Vector2 b) => new(a.X - b.X, a.Y - b.Y);
    
    /// <summary>Multiplies a vector by a scalar.</summary>
    public static Vector2 operator *(Vector2 v, float s) => new(v.X * s, v.Y * s);
    
    /// <summary>Multiplies a scalar by a vector.</summary>
    public static Vector2 operator *(float s, Vector2 v) => new(v.X * s, v.Y * s);

    /// <summary>Calculates the length of the vector.</summary>
    public float Length() => MathF.Sqrt(X * X + Y * Y);

    /// <summary>Returns a normalized copy of the vector.</summary>
    public Vector2 Normalized()
    {
        var len = Length();
        return len > 0 ? new Vector2(X / len, Y / len) : Zero;
    }

    /// <summary>Calculates the dot product with another vector.</summary>
    public float Dot(Vector2 other) => X * other.X + Y * other.Y;

    /// <summary>Zero vector (0, 0).</summary>
    public static Vector2 Zero => new(0, 0);
    
    /// <summary>One vector (1, 1).</summary>
    public static Vector2 One => new(1, 1);

    /// <inheritdoc/>
    public bool Equals(Vector2 other) => X.Equals(other.X) && Y.Equals(other.Y);
    
    /// <inheritdoc/>
    public override bool Equals(object? obj) => obj is Vector2 other && Equals(other);
    
    /// <inheritdoc/>
    public override int GetHashCode() => HashCode.Combine(X, Y);
    
    /// <inheritdoc/>
    public override string ToString() => $"Vector2({X}, {Y})";
}

/// <summary>
/// 3D vector.
/// </summary>
public struct Vector3 : IEquatable<Vector3>
{
    /// <summary>X component.</summary>
    public float X;
    
    /// <summary>Y component.</summary>
    public float Y;
    
    /// <summary>Z component.</summary>
    public float Z;

    /// <summary>
    /// Creates a new 3D vector.
    /// </summary>
    /// <param name="x">X component.</param>
    /// <param name="y">Y component.</param>
    /// <param name="z">Z component.</param>
    public Vector3(float x, float y, float z)
    {
        X = x;
        Y = y;
        Z = z;
    }

    /// <summary>Adds two vectors.</summary>
    public static Vector3 operator +(Vector3 a, Vector3 b) => new(a.X + b.X, a.Y + b.Y, a.Z + b.Z);
    
    /// <summary>Subtracts two vectors.</summary>
    public static Vector3 operator -(Vector3 a, Vector3 b) => new(a.X - b.X, a.Y - b.Y, a.Z - b.Z);
    
    /// <summary>Multiplies a vector by a scalar.</summary>
    public static Vector3 operator *(Vector3 v, float s) => new(v.X * s, v.Y * s, v.Z * s);
    
    /// <summary>Multiplies a scalar by a vector.</summary>
    public static Vector3 operator *(float s, Vector3 v) => new(v.X * s, v.Y * s, v.Z * s);

    /// <summary>Calculates the length of the vector.</summary>
    public float Length() => MathF.Sqrt(X * X + Y * Y + Z * Z);

    /// <summary>Returns a normalized copy of the vector.</summary>
    public Vector3 Normalized()
    {
        var len = Length();
        return len > 0 ? new Vector3(X / len, Y / len, Z / len) : Zero;
    }

    /// <summary>Calculates the dot product with another vector.</summary>
    public float Dot(Vector3 other) => X * other.X + Y * other.Y + Z * other.Z;

    /// <summary>Calculates the cross product with another vector.</summary>
    public Vector3 Cross(Vector3 other) => new(
        Y * other.Z - Z * other.Y,
        Z * other.X - X * other.Z,
        X * other.Y - Y * other.X
    );

    /// <summary>Zero vector (0, 0, 0).</summary>
    public static Vector3 Zero => new(0, 0, 0);
    
    /// <summary>One vector (1, 1, 1).</summary>
    public static Vector3 One => new(1, 1, 1);
    
    /// <summary>Up vector (0, 1, 0).</summary>
    public static Vector3 Up => new(0, 1, 0);
    
    /// <summary>Forward vector (0, 0, -1).</summary>
    public static Vector3 Forward => new(0, 0, -1);
    
    /// <summary>Right vector (1, 0, 0).</summary>
    public static Vector3 Right => new(1, 0, 0);

    /// <inheritdoc/>
    public bool Equals(Vector3 other) => X.Equals(other.X) && Y.Equals(other.Y) && Z.Equals(other.Z);
    
    /// <inheritdoc/>
    public override bool Equals(object? obj) => obj is Vector3 other && Equals(other);
    
    /// <inheritdoc/>
    public override int GetHashCode() => HashCode.Combine(X, Y, Z);
    
    /// <inheritdoc/>
    public override string ToString() => $"Vector3({X}, {Y}, {Z})";
}

/// <summary>
/// 4D vector.
/// </summary>
public struct Vector4 : IEquatable<Vector4>
{
    /// <summary>X component.</summary>
    public float X;
    
    /// <summary>Y component.</summary>
    public float Y;
    
    /// <summary>Z component.</summary>
    public float Z;
    
    /// <summary>W component.</summary>
    public float W;

    /// <summary>
    /// Creates a new 4D vector.
    /// </summary>
    public Vector4(float x, float y, float z, float w)
    {
        X = x;
        Y = y;
        Z = z;
        W = w;
    }

    /// <summary>Zero vector (0, 0, 0, 0).</summary>
    public static Vector4 Zero => new(0, 0, 0, 0);

    /// <inheritdoc/>
    public bool Equals(Vector4 other) => X.Equals(other.X) && Y.Equals(other.Y) && Z.Equals(other.Z) && W.Equals(other.W);
    
    /// <inheritdoc/>
    public override bool Equals(object? obj) => obj is Vector4 other && Equals(other);
    
    /// <inheritdoc/>
    public override int GetHashCode() => HashCode.Combine(X, Y, Z, W);
    
    /// <inheritdoc/>
    public override string ToString() => $"Vector4({X}, {Y}, {Z}, {W})";
}

/// <summary>
/// Quaternion for rotations.
/// </summary>
public struct Quaternion : IEquatable<Quaternion>
{
    /// <summary>X component.</summary>
    public float X;
    
    /// <summary>Y component.</summary>
    public float Y;
    
    /// <summary>Z component.</summary>
    public float Z;
    
    /// <summary>W component.</summary>
    public float W;

    /// <summary>
    /// Creates a new quaternion.
    /// </summary>
    public Quaternion(float x, float y, float z, float w)
    {
        X = x;
        Y = y;
        Z = z;
        W = w;
    }

    /// <summary>Identity quaternion (0, 0, 0, 1).</summary>
    public static Quaternion Identity => new(0, 0, 0, 1);

    /// <inheritdoc/>
    public bool Equals(Quaternion other) => X.Equals(other.X) && Y.Equals(other.Y) && Z.Equals(other.Z) && W.Equals(other.W);
    
    /// <inheritdoc/>
    public override bool Equals(object? obj) => obj is Quaternion other && Equals(other);
    
    /// <inheritdoc/>
    public override int GetHashCode() => HashCode.Combine(X, Y, Z, W);
    
    /// <inheritdoc/>
    public override string ToString() => $"Quaternion({X}, {Y}, {Z}, {W})";
}

File: sdks/go/windjammer/math.go
package windjammer

import "math"

// Vec2 represents a 2D vector
type Vec2 struct {
	X, Y float32
}

// Add adds two vectors
func (v Vec2) Add(other Vec2) Vec2 {
	return Vec2{X: v.X + other.X, Y: v.Y + other.Y}
}

// Sub subtracts two vectors
func (v Vec2) Sub(other Vec2) Vec2 {
	return Vec2{X: v.X - other.X, Y: v.Y - other.Y}
}

// Mul multiplies by a scalar
func (v Vec2) Mul(scalar float32) Vec2 {
	return Vec2{X: v.X * scalar, Y: v.Y * scalar}
}

// Length calculates the length of the vector
func (v Vec2) Length() float32 {
	return float32(math.Sqrt(float64(v.X*v.X + v.Y*v.Y)))
}

// Normalized returns a normalized copy of the vector
func (v Vec2) Normalized() Vec2 {
	length := v.Length()
	if length > 0 {
		return Vec2{X: v.X / length, Y: v.Y / length}
	}
	return Vec2Zero()
}

// Dot calculates the dot product
func (v Vec2) Dot(other Vec2) float32 {
	return v.X*other.X + v.Y*other.Y
}

// Vec2Zero returns a zero vector (0, 0)
func Vec2Zero() Vec2 {
	return Vec2{X: 0, Y: 0}
}

// Vec2One returns a one vector (1, 1)
func Vec2One() Vec2 {
	return Vec2{X: 1, Y: 1}
}

// Vec3 represents a 3D vector
type Vec3 struct {
	X, Y, Z float32
}

// Add adds two vectors
func (v Vec3) Add(other Vec3) Vec3 {
	return Vec3{X: v.X + other.X, Y: v.Y + other.Y, Z: v.Z + other.Z}
}

// Sub subtracts two vectors
func (v Vec3) Sub(other Vec3) Vec3 {
	return Vec3{X: v.X - other.X, Y: v.Y - other.Y, Z: v.Z - other.Z}
}

// Mul multiplies by a scalar
func (v Vec3) Mul(scalar float32) Vec3 {
	return Vec3{X: v.X * scalar, Y: v.Y * scalar, Z: v.Z * scalar}
}

// Length calculates the length of the vector
func (v Vec3) Length() float32 {
	return float32(math.Sqrt(float64(v.X*v.X + v.Y*v.Y + v.Z*v.Z)))
}

// Normalized returns a normalized copy of the vector
func (v Vec3) Normalized() Vec3 {
	length := v.Length()
	if length > 0 {
		return Vec3{X: v.X / length, Y: v.Y / length, Z: v.Z / length}
	}
	return Vec3Zero()
}

// Dot calculates the dot product
func (v Vec3) Dot(other Vec3) float32 {
	return v.X*other.X + v.Y*other.Y + v.Z*other.Z
}

// Cross calculates the cross product
func (v Vec3) Cross(other Vec3) Vec3 {
	return Vec3{
		X: v.Y*other.Z - v.Z*other.Y,
		Y: v.Z*other.X - v.X*other.Z,
		Z: v.X*other.Y - v.Y*other.X,
	}
}

// Vec3Zero returns a zero vector (0, 0, 0)
func Vec3Zero() Vec3 {
	return Vec3{X: 0, Y: 0, Z: 0}
}

// Vec3One returns a one vector (1, 1, 1)
func Vec3One() Vec3 {
	return Vec3{X: 1, Y: 1, Z: 1}
}

// Vec3Up returns the up vector (0, 1, 0)
func Vec3Up() Vec3 {
	return Vec3{X: 0, Y: 1, Z: 0}
}

// Vec3Forward returns the forward vector (0, 0, -1)
func Vec3Forward() Vec3 {
	return Vec3{X: 0, Y: 0, Z: -1}
}

// Vec3Right returns the right vector (1, 0, 0)
func Vec3Right() Vec3 {
	return Vec3{X: 1, Y: 0, Z: 0}
}

